### PR TITLE
Simplify UnitTest.wait and update test suite accordingly

### LINKED ISSUE #5366: Some tests broke due to a change in UnitTest.wait
## Environment

* SuperCollider version: current `develop`

## Issue

A handful of UnitTests broke when `UnitTest.wait` changed in https://github.com/supercollider/supercollider/commit/2aa3374e1e52b7db3f7708498810dc6032a1d082

This breakage is pretty easy to spot. You'll see an error like:
```
^^ ERROR: Message 'test' not understood.
Perhaps you misspelled 'dist', or meant to call 'test' on another receiver?
```

These tests need to be modified so that calls to `UnitTest.wait` are passed a Condition instead of a Function or Boolean.

- TestFilterUGens
- TestNode_Server
- TestServer_boot
- TestMixedBundleTester
- TestCoreUGens
- TestDocument

See discussions in #4659 and #5363
Related issue #5359

### PR DESCRIPTION
**Pull request by jamshark70**
## Purpose and Motivation

Fixes #5366.

After merging #5364, we can simplify `UnitTest:wait` by simply passing the timeout along to the Condition.

This requires all unit tests that had previously passed a function or Boolean now to use a Condition. I've done that here.

It also requires all tests to `.signal` their conditions consistently (which is correct Condition usage anyway). I've updated that.

I reran all of the modified unit tests. All of them except TestCoreUGens pass. I fixed a couple of things in TestCoreUGens but it's still not completely working... after about an hour and a half working on this PR, I'd better put it aside for now.

## Types of changes

- Bug fix
- Breaking change

## To-do list

- [x] Code is tested
- [ ] All tests are passing  -- almost
- [ ] Updated documentation  -- need to?
- [ ] This PR is ready for review  -- almost

### CODE REVIEW

**Comment by brianlheim** [comment]
thanks! how does this PR relate to #3300 and does this fix absolutely require #5364  ? i would *really* prefer to not have to review/merge any new concurrency primitives just to solve this issue, but if you think we ought to do this now, I'm fine reviewing it, just might take a while.

**Comment by brianlheim** [comment]
also if TestCoreUGens wasn't passing before, feel free to leave it for another PR, there is already plenty to review here as is!

**Comment by jamshark70** [comment]
> how does this PR relate to #3300

It looks to me like 3300 isn't entirely valid. It depends on `hang(value)` which will end up double-scheduling threads.

> and does this fix absolutely require #5364 ?

Since this branch is based on the 5364 branch, actually I think it's folded in here.

> i would really prefer to not have to review/merge any new concurrency primitives

There are no new primitives. Timeouts were actually not hard to implement in sclang -- it's just that nobody (including myself) had the right idea before.

#5364 is pretty small, and Julian has already looked at it.

> just to solve this issue, but if you think we ought to do this now

I think this will probably work without #5364 though I didn't test it. But UnitTest:wait is kind of an awkward implementation. If we're fixing the sync mechanism for unit tests, then it would be a good time to improve this as well.

**Comment by jamshark70** [comment]
> how does this PR relate to #3300

It looks to me like 3300 isn't entirely valid. It depends on `hang(value)` which will end up double-scheduling threads.

> and does this fix absolutely require #5364 ?

Since this branch is based on the 5364 branch, actually I think it's folded in here.

> i would really prefer to not have to review/merge any new concurrency primitives

There are no new primitives. Timeouts were actually not hard to implement in sclang -- it's just that nobody (including myself) had the right idea before.

#5364 is pretty small, and Julian has already looked at it.

> just to solve this issue, but if you think we ought to do this now

I think this will probably work without #5364 though I didn't test it. But UnitTest:wait is kind of an awkward implementation. If we're fixing the sync mechanism for unit tests, then it would be a good time to improve this as well.

**Comment by brianlheim** [comment]
> Since this branch is based on the 5364 branch, actually I think it's folded in here.

sorry, yes i understand that -- what i meant was, is it necessary to have 5364 in order to fix #5366?

> There are no new primitives. Timeouts were actually not hard to implement in sclang -- it's just that nobody (including myself) had the right idea before.

yes sorry again, i was being a bit loose with my use of the word "primitive", i meant to indicate that this is kind of a 'fundamental' concurrency operation.

> But UnitTest:wait is kind of an awkward implementation. If we're fixing the sync mechanism for unit tests, then it would be a good time to improve this as well.

i agree the implementation is awkward. ok, so i will look at these two PRs starting with #5364, but just a heads up that it will take me some time to get acquainted with the scheduler code and to think it through. thanks again!

**Comment by jamshark70** [comment]
> is it necessary to have 5364 in order to fix #5366?

It may not be strictly necessary, but I believe the changes here are the correct direction.

> ok, so i will look at these two PRs starting with #5364, but just a heads up that it will take me some time to get acquainted with the scheduler code and to think it through.

Understood. If it will help to ask questions, feel free.

The basic idea is: A timeout is a function that is scheduled in the future, with the provision that a successful result before the timeout expires should cancel the timeout function. I'm doing that here by `Routine { timeout.wait; timeout stuff here }`. If the condition doesn't release (it times out), then the routine awakens and does the timeout stuff. If it the condition does release (`unhang`, or `signal` when the condition is true), then we `stop` the Routine. The routine is still on the clock, but when it awakens, its state prevents the timeout stuff from running. So you get either a successful release or a timeout, but never both.

Routine just serves here as a cancel-able function wrapper... not the normal usage of Routine (which is probably why it was elusive) but I think it's valid. It passes every test I can think of, at least.

**Review by telephon — COMMENTED** [review]
Running all the tests, I got only two failures:

```
UNIT TESTS FOR 'TestDocument' COMPLETED
There were failures:
a TestDocument: test_new_document_runs_initAction - Document.new should fire Document.initAction function

FAIL: a TestDocument: test_new_document_runs_initAction



UNIT TESTS FOR 'TestFilterUGens' COMPLETED
There were failures:
a TestFilterUGens: test_time_invariance

```

In current `develop`, both fail with `ERROR: Message 'test' not understood.`

File: testsuite/classlibrary/TestDocument.sc [hunk 1]
```diff
@@ -5,14 +5,15 @@ TestDocument : UnitTest {
 
 	test_new_document_runs_initAction {
 		var success = false, doc, save;
```
Comment (patrickdupuis) [inline testsuite/classlibrary/TestDocument.sc:7]: The `success` variable isn't needed anymore, and it shouldn't be used on line 21 in the assert.

**Comment by patrickdupuis** [comment]
Strange, `TestDocument.test_new_document_runs_initAction` reports both `PASS` and `FAIL` when making the changes I commented above :thinking: 

```
RUNNING UNIT TEST 'TestDocument'
-> TestDocument
PASS: a TestDocument: test_new_document_runs_initAction - Document.new should fire Document.initAction function
PASS: a TestDocument: test_open_autorun_document_runs_code - Document.open should run autoRun /*RUN*/ documents
PASS: a TestDocument: test_document_getText_retrievesText - Document tests are valid only when run in the IDE
PASS: a TestDocument: test_document_getText_retrievesText - getText contents retrieved from document should match input contents
Is:
	 abc 
Should be:
	 abc
PASS: a TestDocument: test_document_getTextAsync_retrievesText - Document tests are valid only when run in the IDE
PASS: a TestDocument: test_document_getTextAsync_retrievesText - getTextAsync contents retrieved from document should match input contents
Is:
	 abc 
Should be:
	 abc

UNIT TESTS FOR 'TestDocument' COMPLETED
There were no failures

FAIL: a TestDocument: test_new_document_runs_initAction
```

**Comment by jamshark70** [comment]
> Strange, TestDocument.test_new_document_runs_initAction reports both PASS and FAIL when making the changes I commented above

OK, fixed. Good catch.

The problem is that Document.new runs its initAction synchronously.

The desired sequence of events is:

1. Set up initAction and make the document.
2. Wait with timeout.
3. initAction fires and hangs the thread.

The actual sequence of events was:

1. Set up initAction and make the document.
2. initAction fires immediately and signals the thread to continue.
3. *Then* it sets the timeout, and `wait` doesn't wait because the condition is already true.

But the condition was already signaled, so there's nothing to stop the timeout.

That means we *do* need `hangWithTimeout` and `waitWithTimeout` in Condition after all -- because `timeoutAfter` doesn't know whether the thread will be paused by force (`hang`) or conditionally (`wait`). `waitWithTimeout` should set the timeout only if the condition is currently false.

With this fix, the sequence is: 1. Doc.new --> initAction; 2. Condition is already true, so it skips the timeout and the wait; 3. assert.

This does call for a documentation update but I don't have time just this second to do it. Will get to it later this morning.

**Comment by dyfer** [comment]
Open question - do you all think it would be desirable to include in this PR enabling previously disabled tests in CI, for the tests that are fixed here? That would entail removing appropriate entries in [gha_test_run_proto.json](https://github.com/supercollider/supercollider/blob/develop/testsuite/scripts/gha_test_run_proto.json)

**Comment by jamshark70** [comment]
@brianlheim

> also if TestCoreUGens wasn't passing before, feel free to leave it for another PR

I just see now, according to https://github.com/supercollider/supercollider/blob/develop/testsuite/scripts/gha_test_run_proto.json, "UnitTest has bugs (FIXME)" for core UGens so the failures I saw after updating are likely not regressions.

So this PR would suggest a canonical way to do sync in unit tests, which future revisions to the core UGens test would take into account, but it isn't fixing old bugs.

@dyfer Agreed, any tests that were broken before but which are passing after these changes should be enabled, after merging this.

**Comment by patrickdupuis** [comment]
`TestMovingSum` throws an error `Message waitWithTimeout not understood`

**Comment by patrickdupuis** [comment]
Sorry, `TestMovingSum` is part of sc3-plugins. Ignore what I just said.

**Comment by patrickdupuis** [comment]
I just finished running all our UnitTests againsts this PR. Overall, I didn't see any new breakage compared to `develop`. However, I would suggest we look into `TestLinkClock` and `TestProxyNodeMap` to make sure they haven't broken.

**Comment by telephon** [comment]
`TestProxyNodeMap` makes no use of `wait`.

**Comment by patrickdupuis** [comment]
I've checked `TestProxyNodeMap` and `TestLinkClock`. They both have failures on `develop` which are also present here. I thought maybe the failures were new to this branch, but that's not the case.

I've gone through all the class library tests and haven't come accross any new failures. So, this PR is looking good to me.

**Comment by telephon** [comment]
For now, `test_nodeMap_controlNamesConvertsObjectToControlInput` can be commented out. I don't know anymore what I was thinking then. I need to check this separately.

**Comment by telephon** [comment]
The `TestLinkClock` methods all run here (macOS) but need to have the right permissisons set first. Also they take quite a long time (around 3 sec per test).

**Comment by dyfer** [comment]
> For now, `test_nodeMap_controlNamesConvertsObjectToControlInput` can be commented out. I don't know anymore what I was thinking then. I need to check this separately.

@telephon please see https://github.com/supercollider/supercollider/wiki/GitHub-Actions-migration-notes#test-suite for information which tests are being run. Right now all `TestProxyNodeMap` are [disabled](https://github.com/supercollider/supercollider/blob/develop/testsuite/scripts/gha_test_run_proto.json#L145), is `test_nodeMap_controlNamesConvertsObjectToControlInput` the only one that's failing? If that's the case, it could be excluded specifically.

**Comment by telephon** [comment]
thanks!

**Comment by brianlheim** [comment]
@jamshark70 just a heads-up that I have been reviewing this, but it's taking a while because I want to make sure I understand everything going on! I'll post something tomorrow or Sunday. thanks for your patience.

**Comment by jamshark70** [comment]
> just a heads-up that I have been reviewing this, but it's taking a while...

Ok, great!

I just spotted a couple of things to change, but I'll wait for your review before doing it:

- `waitWithTimeout` should probably call `this.hang` instead of `this.wait`. As it is now, both `waitWithTimeout` and `wait` evaluate the test function -- double evaluation. We don't expect the test function to have side effects, but if it does, they will be repeated an extra time. Instead, `waitWithTimeout` should evaluate it and, if the test is false, just force-hang without asking again.

- `timeoutAfter` should probably reverse the order of `action.value;` and `waitingThread.clock.sched(0, resume);` -- as it is now, an error in the action function would fail to resume the thread. (Or do we want it to fail to resume, on error?)

**Comment by brianlheim** [comment]
Thanks for this @jamshark70 , sorry it took a while to review and sorry I have generally
not been participating that much in these discussions, I know it is a major pain point for
you with testing.

My opinion after taking some time to think about this and reviewing condition variables in
other libraries is that Condition in SC has too many defects to be workable, and that we
should instead put effort into designing a new object with a simpler and easier to use
interface (which I will tentatively call `CondVar` to avoid confusion). what I'm
suggesting combines ideas from both this PR and from #3895, would probably go through the
RFC process, and most likely target the 3.13.0 release, although it could be done in time
for 3.12.0.

Also for clarity's sake, below I'm going to refer to the typically Function-like object
that's held by Condition:-test as the **predicate** rather than **test** because the
latter has overloaded meaning.

I'm going to share what I see as the major flaws of Condition, and then what I'd suggest
as the new interface to replace it.

#### 1. The predicate is given to the constructor and is held by the class

Declaring the predicate when constructing the Condition variable is inconvenient. Also,
having just one predicate which is managed by the class is an unnecessary restriction and
introduces state where none is necessary.  Instead, the predicate should be passed to
`wait` itself as an optional parameter; or, if desired, there can be separate methods for
waiting and for waiting-with-predicate, a bit like how `hang` and `wait` are right now.
Personally, I would combine them to a single function, and this is what I suggested below.

Some issues with a Condition with stateful predicate:
- it's confusing that Condition takes a `test` parameter in its constructor, but that it
  isn't always used (in hang and unhang, for example).
- it cannot support the cases where multiple waiting threads want to consume the same
  resource in differing ways or degrees, and therefore have different predicates.
- everything can be managed without passing any predicates to Condition at all, and in
  fact this is what many Condition designs do.
- the predicate may depend on variables/state available long after the Condition has been
  constructed, requiring users to make an unnecessary choice between rearranging code to
  fit this requirement, or setting `cond.test` later on, which may lead to confusing bugs
  regarding state manipulation.
- it unnecessarily separates the predicate code from the callsite of `wait`, when we
  usually care about both together.
- the predicate is not guaranteed to always be evaluated on the same thread.

For comparison, here is the same bit of code using Condition, and then a hypothetical
CondVar which supports wait-with-predicate:

<details>
<summary>Code</summary>

```
(
var x = 0;
var c = Condition { x.postln == 6 }; // must be declared after x!

/* lots of code */

// have to jump up to see the predicate; no guarantee on what thread the predicate is
// evaluated
fork { c.wait; "done".postln; };
fork { loop { x = x + 1; c.signal; 0.25.wait; } };
)

(
var c = Condition(); // predicate set below
var x = 0; // can now be declared after `c`

/* lots of code */

// set predicate as close as possible to `wait` callsite; risky because we may be
// overwriting `c.test` without knowing it, and because some later code may overwrite it.
c.test = { x.postln == 6 };
fork { c.wait; "done".postln; };
fork { loop { x = x + 1; c.signal; 0.25.wait; } };
)

(
var c = CondVar(); // can declare before or after `x`
var x = 0;

/* lots of code */

// predicate is immediately clear, and always evaluated on the thread calling `wait`
fork { c.wait { x.postln == 6 }; "I have 6".postln };
// we can even have multiple waits with distinct predicates
fork { c.wait { x.postln == 5 }; "I have 5".postln };
fork { loop { x = x + 1; c.signal; 0.25.wait; } };
)
```

</details>

#### 2. The hang/wait and signal/unhang distinctions break the abstraction of the class

Condition not only supports two orthogonal interfaces at once, but it also allows them to
be mixed with bad results. `unhang` can wake `wait`ing threads even when `test.value` is
`false`, and `signal` will only wake `hang`ing threads if `test.value` is `true`. This
behavior is undesirable because there is no way to tell after a call to `wait` returns
whether or not the predicate is true (or tested true when the thread was woken). It is
also confusing for users to learn the difference between these two interfaces, when things
can actually be much simpler and more consistent with a single interface. Instead of both
`signal` and `unhang`, there should be a single method, `signal` (or `notify`, typically
used for other implementations) which wakes threads, and then the predicate test is run on
each thread independently and the decision to continue blocking is made there. Instead of
both `wait` and `hang`, there should be a single method `wait`, which has a single
optional predicate parameter. With this interface, it is possible to guarantee that
a thread returns from `wait` if and only if the condition was signaled and the predicate,
if one was passed, is true.

For comparison:

<details>
<summary>Code</summary>

```
(
c = Condition(); // `test` defaults to `false`

// first thread will never be woken, as by default `c.test === false`
fork { c.hang; "hang done".postln; };
fork { 1.wait; c.signal; };

// first thread will be woken, even though `c.test === false`
fork { c.wait; "wait done".postln; };
fork { 1.wait; c.unhang; };

// looks correct, but misuses interface -- `unhang` does not depend on `cond.test.value`
fork { c.wait; "second wait done".postln; };
fork { 1.wait; c.test = true; c.unhang; };

// is correct, but strange use of interface -- return from `hang` deceptively depends on `cond.test.value`
fork { c.hang; "second wait done".postln; };
fork { 1.wait; c.test = true; c.signal; };
)
```

</details>

#### 3. There is no easy way to wake up a single waiting/hanging thread

All implementations of Condition I could find support waking a single thread as opposed to
waking all threads. This is desirable whenever we have multiple threads waiting on a
single condition and we know only one will be able to use the resource guarded by the
condition next. So, CondVar should support both `notifyOne` and `notifyAll`. The verb
`signal` can also be used instead of `notify`; this is just the most common name I see for
the operation.

#### Proposed new interface

<details>
<summary>Here is the interface I think we should approach for CondVar.</summary>

```
/* Creates a CondVar.
 */
new()

/* If `pred` is `nil` (its default value), blocks the current thread until woken by
 * `notifyOne()` or `notifyAll()`. Otherwise, if `pred.value` returns true, returns
 * immediately.  Otherwise, blocks the current thread until woken by `notifyOne()` or
 * `notifyAll()` and `pred.value` returns `true`. `pred` is typically a Boolean or Function.
 * If `pred` throws an exception, the exception will be propagated up on the thread which
 * called `wait`, and the thread which called it will no longer be considered waiting on
 * this CondVar.
 *
 * With a predicate, this method is equivalent to:
 *
 *     while { pred.value.not } { cond.wait }
 *
 * Returns this object.
 */
wait(pred)

/* Wakes one thread waiting on this Condition. Threads are woken in the order in which
 * they called `wait()`. If a thread is woken and was waiting with a predicate, and that
 * predicate is still `false`, it will `wait` again and be placed at the end of the queue
 * of threads waiting on this CondVar. Another thread will not be woken in that case.
 * Returns this object.
 */
notifyOne()

/* Similar to `notifyOne()` but all threads waiting on this CondVar are woken.
 * Returns this object.
notifyAll()

/* If `pred` is not nil and `pred.value` returns true, returns immediately. Otherwise,
 * blocks the current thread until either the thread is woken by `notifyOne()` or
 * `notifyAll()` and `pred.value` returns `true`, or `beats` time in clock beats has
 * elapsed. Returns `false` if the timeout elapsed and either `pred` was `nil` or
 * `pred.value` was `false`, otherwise `true`.  In other words, this method behaves much like
 * `wait` except that if it returns `false`, the wait timed out without a passed predicate
 * becoming true.

 * If `pred` throws an exception, the exception will be propagated up on the thread which
 * called `waitWithTimeout`, and the thread which called it will no longer be considered
 * waiting on this CondVar.
 *
 * With a predicate, this is equivalent to:
 *
 *     var beatsRemaining = beats;
 *     while { pred.value.not } {
 *         if(cond.waitWithTimeout(beatsRemaining).not) {
 *             ^pred.value
 *         };
 *         // Very generalized pseudocode -- necessary because this is
 *         // wait-for-duration and not wait-until-timepoint.
 *         beatsRemaining = calculateBeatsRemaining(thisThread.threadPlayer, beats);
 *     }
 *     ^true
 *
 * If `beats` is not a Float or Integer, an exception is immediately thrown.
 */
waitWithTimeout(beats, pred)

/* Throws an error; CondVar is not shallow-copyable or copyable.
 */
shallowCopy()

/* Throws an error; CondVar is not deep-copyable.
 */
deepCopy()
```

</details>

#### Exception safety

I also haven't used an `action` callback like in your `timeoutAfter` design; this is
because I think the same thing can be achieved with greater simplicity by returning a
Boolean indicating whether the timeout expired without an optional predicate becoming
true. Adding a timeout-expired callback requires us to think about (1) what happens if an
exception is thrown in that callback, and (2) which thread the callback should run on.

For comparison:

<details>
<summary>Code</summary>

```
(
c = Condition();

// To what does `thisThread` refer? How do we catch the exception?
c.waitWithTimeout(3) { thisThread.hash.postln; Exception("whoops!").throw };

// `thisThread` refers to the thread calling `waitWithTimeout`; the exception will
// propagate on this thread's stack as usual.
if(c.waitWithTimeout(3).not) { thisThread.hash.postln; Exception("whoops!").throw };
)
```

</details>

I'm also intentionally suggesting that an error be thrown on the calling thread if beats
isn't a Float or Integer, because it's much simpler to manage an exception being thrown
that way than deep within a forked thread. I'm not sure if the implementation will allow
this but ideally it should be almost impossible for an exception to be thrown while
"context switching", because this just creates massive headaches for anyone trying to
understand the code or debug an error they hit.

---

Initially, I was thinking it might be possible to get to this interface through a series
of deprecations. However, this would be really quite tricky to do without any mistakes,
and if we are going to start recommending Condition to more people, my view is that it
would be better to not burden them with documentation that has to explain both the old and
new interfaces. I would rather just deprecate the entire class and start with a fresh one.

Sorry if this seems like a lot; I know you were kind of expecting this to be merged as-is,
but I really don't think this is a good idea. For such a fundamental class, I really think
think we need to get the abstraction right, and right now I just feel there are too many
issues with Condition as it's currently implemented.

Let me know how you'd like to proceed. If you think this looks all right, maybe we could
proceed with a new PR. However, I feel like it may be better to go to an RFC, which I'd be
happy to initiate if you don't feel you have enough time to manage it. Also, if this is
going to take a while, we should actually probably add a quick band-aid so that these
failing tests work again.

**Comment by jamshark70** [comment]
As for the technical points -- I can't object. A new class would be better.

I have some thoughts about deprecation (the current mechanism removes an element of consent from the user) and about review process, which don't belong here.

My semester is beginning and my free time to manage PRs is about to disappear, so I'm afraid I won't be able to continue with this topic in the immediate future. Also I'm a bit gun-shy to write it myself, if you have something specific in mind for the implementation. I'll keep my eyes open for a future PR about this.

**Comment by brianlheim** [comment]
okay, I'll close this PR then and submit a new PR in the next week or so. Thanks!

**Pull request closed without merging**